File: knowledge/tools/vitali87-code-graph-rag/evals/oracles/java_oracle/Oracle.java
// Authoritative Java structure oracle for the cgr eval harness.
//
// Parses every .java file under a directory with the JDK's own Compiler Tree API
// (javax.tools + com.sun.source) and emits one JSON record per declaration, in
// cgr's NodeLabel vocabulary, so records join cgr's graph on (kind, file, line).
// task.parse() only parses (no resolution), so missing dependencies are fine.
//
// Mapping (Java construct -> cgr NodeLabel):
//
//   class                  -> Class
//   interface / @interface -> Interface  (its method signatures -> Method)
//   enum                   -> Enum
//   method / constructor   -> Method
//
// Containment edges (matching how cgr models Java containment):
//
//   DEFINES        : the file module -> every named type (top-level OR nested)
//   DEFINES_METHOD : the method's immediate enclosing named type -> Method
//
// cgr models a Java module per file (keyed at line 0) and DEFINES every named
// type from it (containment is flat, not nested-type-scoped). A method binds to
// its nearest enclosing named type. Methods of an anonymous class are Functions
// (no DEFINES_METHOD), matching the node mapping.
//
// Output is a {nodes, edges} payload joining cgr on (kind, file, line).
//
// Compile: javac Oracle.java ; Run: java -cp <dir> Oracle <dir>

import com.sun.source.tree.ClassTree;
import com.sun.source.tree.CompilationUnitTree;
import com.sun.source.tree.ExpressionTree;
import com.sun.source.tree.IdentifierTree;
import com.sun.source.tree.LambdaExpressionTree;
import com.sun.source.tree.LineMap;
import com.sun.source.tree.MemberSelectTree;
import com.sun.source.tree.MethodInvocationTree;
import com.sun.source.tree.MethodTree;
import com.sun.source.tree.Tree;
import com.sun.source.util.JavacTask;
import com.sun.source.util.SourcePositions;
import com.sun.source.util.TreePath;
import com.sun.source.util.TreePathScanner;
import com.sun.source.util.Trees;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import javax.tools.JavaCompiler;
import javax.tools.JavaFileObject;
import javax.tools.StandardJavaFileManager;
import javax.tools.ToolProvider;

public class Oracle {
    static final Set<String> IGNORED =
        new HashSet<>(Arrays.asList(".git", "target", "build", "node_modules", "vendor"));
    static final List<String> recs = new ArrayList<>();
    static final List<String> edges = new ArrayList<>();
    static final List<String> nameEdges = new ArrayList<>();
    static final List<String> calls = new ArrayList<>();
    static final long MODULE_LINE = 0;

    static String esc(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    // Simple name of an extends/implements type: drop generics and any
    // package/outer qualifier, matching how cgr resolves bases by simple name.
    static String simpleName(Object typeTree) {
        String s = typeTree.toString();
        int lt = s.indexOf('<');
        if (lt >= 0) {
            s = s.substring(0, lt);
        }
        int dot = s.lastIndexOf('.');
        if (dot >= 0) {
            s = s.substring(dot + 1);
        }
        return s.trim();
    }

    static void emitNameEdge(
            String rel, String file, String skind, long sline, String targetName) {
        nameEdges.add("{\"rel\":\"" + rel + "\",\"source\":{\"kind\":\"" + skind
            + "\",\"file\":\"" + esc(file) + "\",\"line\":" + sline
            + "},\"target_name\":\"" + esc(targetName) + "\"}");
    }

    // A file-level call site: caller file + callee simple name (the method
    // identifier). The Python side keeps only callees whose name is a declared
    // first-party Method/Function, mirroring the Go/Rust call oracles.
    static void emitCall(String file, String name) {
        calls.add("{\"file\":\"" + esc(file) + "\",\"name\":\"" + esc(name) + "\"}");
    }

    static void emit(String kind, String file, long line, long endLine, String name) {
        recs.add("{\"kind\":\"" + kind + "\",\"file\":\"" + esc(file)
            + "\",\"line\":" + line + ",\"end_line\":" + endLine
            + ",\"name\":\"" + esc(name) + "\"}");
    }

    static void emitEdge(
            String rel, String file, String pkind, long pline, String ckind, long cline) {
        edges.add("{\"rel\":\"" + rel + "\",\"parent\":{\"kind\":\"" + pkind
            + "\",\"file\":\"" + esc(file) + "\",\"line\":" + pline
            + "},\"child\":{\"kind\":\"" + ckind + "\",\"file\":\"" + esc(file)
            + "\",\"line\":" + cline + "}}");
    }

    static String classKind(ClassTree node) {
        switch (node.getKind()) {
            case INTERFACE:
                return "Interface";
            case ENUM:
                return "Enum";
            // cgr models an annotation type (@interface) as a Class.
            default:
                return "Class";
        }
    }

    public static void main(String[] args) throws Exception {
        Path root = Paths.get(args[0]).toAbsolutePath().normalize();
        List<Path> files = new ArrayList<>();
        Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
            public FileVisitResult preVisitDirectory(Path d, BasicFileAttributes a) {
                Path name = d.getFileName();
                if (name != null && IGNORED.contains(name.toString())) {
                    return FileVisitResult.SKIP_SUBTREE;
                }
                return FileVisitResult.CONTINUE;
            }

            public FileVisitResult visitFile(Path f, BasicFileAttributes a) {
                if (f.toString().endsWith(".java")) {
                    files.add(f);
                }
                return FileVisitResult.CONTINUE;
            }
        });
        if (files.isEmpty()) {
            System.out.print("{\"nodes\":[],\"edges\":[],\"name_edges\":[]}");
            return;
        }

        JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
        StandardJavaFileManager fm = compiler.getStandardFileManager(null, null, null);
        Iterable<? extends JavaFileObject> units = fm.getJavaFileObjectsFromPaths(files);
        JavacTask task = (JavacTask) compiler.getTask(null, fm, d -> {}, null, null, units);
        SourcePositions sp = Trees.instance(task).getSourcePositions();

        for (CompilationUnitTree unit : task.parse()) {
            Path abs = Paths.get(unit.getSourceFile().toUri());
            String rel = root.relativize(abs).toString().replace('\\', '/');
            LineMap lm = unit.getLineMap();
            new TreePathScanner<Void, Void>() {
                public Void visitClass(ClassTree node, Void p) {
                    long pos = sp.getStartPosition(unit, node);
                    // Anonymous classes have an empty name and no cgr node.
                    if (pos >= 0 && node.getSimpleName().length() > 0) {
                        long line = lm.getLineNumber(pos);
                        long endLine = lm.getLineNumber(sp.getEndPosition(unit, node));
                        String kind = classKind(node);
                        emit(kind, rel, line, endLine, node.getSimpleName().toString());
                        // Every named type is DEFINEd by the file module,
                        // including nested types (cgr keeps this flat).
                        emitEdge("DEFINES", rel, "Module", MODULE_LINE, kind, line);
                        // extends superclass -> INHERITS (a class only).
                        if (node.getExtendsClause() != null) {
                            emitNameEdge("INHERITS", rel, kind, line,
                                simpleName(node.getExtendsClause()));
                        }
                        // The implements clause holds a class/enum's interfaces
                        // (-> IMPLEMENTS) but an interface's superinterfaces
                        // (-> INHERITS, like cgr).
                        String hrel = node.getKind() == Tree.Kind.INTERFACE
                            ? "INHERITS" : "IMPLEMENTS";
                        for (Tree it : node.getImplementsClause()) {
                            emitNameEdge(hrel, rel, kind, line, simpleName(it));
                        }
                    }
                    return super.visitClass(node, p);
                }

                public Void visitMethod(MethodTree node, Void p) {
                    long pos = sp.getStartPosition(unit, node);
                    if (pos >= 0) {
                        // cgr labels a member a Method only when its nearest
                        // enclosing named class precedes any enclosing method or
                        // lambda body; members of an anonymous class (declared in
                        // a method body) are modelled as standalone Functions.
                        String kind = "Function";
                        ClassTree owner = null;
                        for (TreePath up = getCurrentPath().getParentPath();
                                up != null; up = up.getParentPath()) {
                            Tree t = up.getLeaf();
                            if (t instanceof ClassTree
                                    && ((ClassTree) t).getSimpleName().length() > 0) {
                                kind = "Method";
                                owner = (ClassTree) t;
                                break;
                            }
                            if (t instanceof MethodTree || t instanceof LambdaExpressionTree) {
                                break;
                            }
                        }
                        long line = lm.getLineNumber(pos);
                        long endLine = lm.getLineNumber(sp.getEndPosition(unit, node));
                        emit(kind, rel, line, endLine, node.getName().toString());
                        // A Method binds to its enclosing named type; an
                        // anonymous-class member (Function) has no owner node,
                        // so cgr anchors it to the module with DEFINES (the
                        // orphan-prevention fallback) and the oracle models
                        // the same containment.
                        if (owner != null) {
                            long opos = sp.getStartPosition(unit, owner);
                            if (opos >= 0) {
                                emitEdge("DEFINES_METHOD", rel, classKind(owner),
                                    lm.getLineNumber(opos), "Method", line);
                            }
                        } else {
                            emitEdge("DEFINES", rel, "Module", MODULE_LINE,
                                "Function", line);
                        }
                    }
                    return super.visitMethod(node, p);
                }

                public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
                    // The callee simple name: the trailing identifier of a
                    // member-select (`obj.foo()`, `Type.bar()`) or a bare
                    // identifier (`foo()`, same-class or static-imported). A
                    // `super()`/`this()` constructor call yields "super"/"this"
                    // and is dropped downstream (never a declared method name).
                    ExpressionTree sel = node.getMethodSelect();
                    String name = null;
                    if (sel instanceof MemberSelectTree) {
                        name = ((MemberSelectTree) sel).getIdentifier().toString();
                    } else if (sel instanceof IdentifierTree) {
                        name = ((IdentifierTree) sel).getName().toString();
                    }
                    if (name != null) {
                        emitCall(rel, name);
                    }
                    return super.visitMethodInvocation(node, p);
                }
            }.scan(unit, null);
        }
        System.out.print("{\"nodes\":[" + String.join(",", recs)
            + "],\"edges\":[" + String.join(",", edges)
            + "],\"name_edges\":[" + String.join(",", nameEdges)
            + "],\"calls\":[" + String.join(",", calls) + "]}");
    }
}

File: knowledge/tools/vitali87-code-graph-rag/evals/oracles/scala_oracle/Oracle.scala
//> using scala 2.13.14
//> using dep org.scalameta::scalameta:4.9.9

// Independent Scala call oracle for cgr's retrieval eval. Parses each .scala
// source with scalameta (NOT tree-sitter, which cgr uses) and emits one JSON
// payload: `nodes` (every def declaration, kind Function/Method + name) and
// `calls` (each call site as relative-file + callee simple name). The retrieval
// driver restricts calls to first-party names (a callee whose simple name is a
// declared def) and grades cgr's Scala CALLS against these sites.
import scala.meta._
import java.nio.file.{Files, Path, Paths, FileVisitResult, SimpleFileVisitor}
import java.nio.file.attribute.BasicFileAttributes
import scala.collection.mutable.ArrayBuffer

object Oracle {
  val IGNORED: Set[String] =
    Set(".git", ".venv", "venv", "node_modules", "target", "__pycache__", ".mypy_cache")

  val nodes = ArrayBuffer.empty[String]
  val calls = ArrayBuffer.empty[String]
  val covered = ArrayBuffer.empty[String]

  def esc(s: String): String =
    s.flatMap {
      case '"'  => "\\\""
      case '\\' => "\\\\"
      case '\n' => "\\n"
      case '\r' => "\\r"
      case '\t' => "\\t"
      case c    => c.toString
    }

  def emitNode(kind: String, file: String, line: Int, endLine: Int, name: String): Unit =
    nodes += s"""{"kind":"${esc(kind)}","file":"${esc(file)}","line":$line,"end_line":$endLine,"name":"${esc(name)}"}"""

  def emitCall(file: String, name: String): Unit =
    calls += s"""{"file":"${esc(file)}","name":"${esc(name)}"}"""

  // The trailing simple name of a callee expression: `foo`, `obj.foo`,
  // `pkg.Obj.foo`, `foo[T]`, curried `foo(a)(b)` all reduce to `foo`.
  def trailingName(fun: Term): Option[String] = fun match {
    case Term.Name(v)            => Some(v)
    case Term.Select(_, name)    => Some(name.value)
    case Term.ApplyType(f, _)    => trailingName(f)
    case Term.Apply(f, _)        => trailingName(f)
    case _                       => None
  }

  // Nearest enclosing template (class/trait/object body) => Method, else Function.
  def process(file: String, tree: Tree): Unit = {
    def walk(t: Tree, inTemplate: Boolean): Unit = {
      t match {
        // Skip package refs and imports: their Term.Select nodes are not calls.
        case Pkg(_, stats) =>
          stats.foreach(walk(_, inTemplate)); return
        case _: Import => return
        case _ =>
      }

      t match {
        case d: Defn.Def =>
          val kind = if (inTemplate) "Method" else "Function"
          emitNode(kind, file, d.pos.startLine + 1, d.pos.endLine + 1, d.name.value)
        case d: Decl.Def =>
          val kind = if (inTemplate) "Method" else "Function"
          emitNode(kind, file, d.pos.startLine + 1, d.pos.endLine + 1, d.name.value)
        case Term.Apply(fun, _)         => trailingName(fun).foreach(emitCall(file, _))
        case Term.ApplyInfix(_, op, _, _) => emitCall(file, op.value)
        case Term.ApplyType(fun, _)     => trailingName(fun).foreach(emitCall(file, _))
        // A bare `x.name` with no application (a standalone Term.Select) is NOT
        // emitted as a call: Scala's uniform access makes a nullary method call and
        // a plain field read syntactically identical, so counting it would treat a
        // same-named `val` read as a first-party call. The eval therefore grades
        // application/infix call sites only, matching cgr's Scala call extraction,
        // which for the same reason does not resolve bare `field_expression` nodes.
        case _ =>
      }

      val childTemplate = t.isInstanceOf[Template] || inTemplate
      t.children.foreach(walk(_, childTemplate))
    }
    walk(tree, inTemplate = false)
  }

  def main(args: Array[String]): Unit = {
    val root = Paths.get(args(0)).toAbsolutePath.normalize
    val files = ArrayBuffer.empty[Path]
    Files.walkFileTree(root, new SimpleFileVisitor[Path] {
      override def preVisitDirectory(d: Path, a: BasicFileAttributes): FileVisitResult = {
        val name = d.getFileName
        if (name != null && IGNORED.contains(name.toString)) FileVisitResult.SKIP_SUBTREE
        else FileVisitResult.CONTINUE
      }
      override def visitFile(f: Path, a: BasicFileAttributes): FileVisitResult = {
        val s = f.toString
        if (s.endsWith(".scala") || s.endsWith(".sc")) files += f
        FileVisitResult.CONTINUE
      }
    })

    files.foreach { f =>
      val rel = root.relativize(f).toString.replace('\\', '/')
      val text = new String(Files.readAllBytes(f), "UTF-8")
      Input.VirtualFile(rel, text).parse[Source] match {
        case Parsed.Success(tree) =>
          covered += s""""${esc(rel)}""""
          process(rel, tree)
        case Parsed.Error(_, _, _) => // skip unparseable file (oracle grades only clean parses)
      }
    }

    print(s"""{"nodes":[${nodes.mkString(",")}],"edges":[],"name_edges":[],"calls":[${calls.mkString(",")}],"covered":[${covered.mkString(",")}]}""")
  }
}
